src/lib.rs

```diff
@@ -1,14 +1,331 @@
+    fn from(slice: &'a [u8]) -> Self {
+        Self(Der::from(slice))
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl<'a> From<Vec<u8>> for CertificateRevocationListDer<'a> {
+    fn from(vec: Vec<u8>) -> Self {
+        Self(Der::from(vec))
+    }
+}
+
+/// A DER-encoded X.509 certificate; as specified in RFC 5280
+///
+/// A common format for storing certificates is
+/// [PEM](https://en.wikipedia.org/wiki/Privacy-Enhanced_Mail).
+/// PEM certificates are commonly stored in files with a `.pem`, `.cer` or `.crt` suffix, and look
+/// like this:
+///
+/// ```txt
+/// -----BEGIN BEGIN EC PRIVATE KEY-----
+/// <base64-encoded certificate content>
+/// -----END CERTIFICATE-----
+/// ```
+///
+/// The [`rustls-pemfile`](https://docs.rs/rustls-pemfile/latest/rustls_pemfile/) crate can be used
+/// to parse PEM files. The [`rcgen`](https://docs.rs/rcgen/latest/rcgen/) crate can be used to
+/// generate certificates and private keys.
+/// ```
+#[derive(Clone, Debug, PartialEq)]
+pub struct CertificateDer<'a>(Der<'a>);
+
+impl AsRef<[u8]> for CertificateDer<'_> {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl Deref for CertificateDer<'_> {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
+impl<'a> From<&'a [u8]> for CertificateDer<'a> {
+    fn from(slice: &'a [u8]) -> Self {
+        Self(Der::from(slice))
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl<'a> From<Vec<u8>> for CertificateDer<'a> {
+    fn from(vec: Vec<u8>) -> Self {
+        Self(Der::from(vec))
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub struct Der<'a>(DerInner<'a>);
+
+impl<'a> Der<'a> {
+    // Trait methods cannot be const, so we additionally offer some traditional constructors.
+    pub const fn from_slice(der: &'a [u8]) -> Self {
+        Self(DerInner::Slice(der))
+    }
+}
+
+impl AsRef<[u8]> for Der<'_> {
+    fn as_ref(&self) -> &[u8] {
+        match &self.0 {
+            #[cfg(feature = "alloc")]
+            DerInner::Vec(vec) => vec.as_ref(),
+            DerInner::Slice(slice) => slice,
+        }
+    }
+}
 
-    #[test]
-    fn it_works() {
-        let result = add(2, 2);
-        assert_eq!(result, 4);
+impl Deref for Der<'_> {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl<'a> From<&'a [u8]> for Der<'a> {
+    fn from(slice: &'a [u8]) -> Self {
+        Self(DerInner::Slice(slice))
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl<'a> From<Vec<u8>> for Der<'a> {
+    fn from(vec: Vec<u8>) -> Self {
+        Self(DerInner::Vec(vec))
+    }
+}
+
+impl fmt::Debug for Der<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("Der").field(&self.as_ref()).finish()
     }
 }
+
+#[derive(Clone, PartialEq)]
+enum DerInner<'a> {
+    #[cfg(feature = "alloc")]
+    Vec(Vec<u8>),
+    Slice(&'a [u8]),
```

Comment:
I would have called these variants `Owned` and `Borrowed` but both work for me.

Reply:
Makes sense, I can change them.